db/dcrpg/internal/stakestmts.go

```diff
@@ -0,0 +1,218 @@
+	insertTicketRowReturnId = `WITH ins AS (` +
+		insertTicketRow0 +
+		`ON CONFLICT (tx_hash, block_hash) DO UPDATE
+		SET tx_hash = NULL WHERE FALSE
+		RETURNING id
+		)
+	SELECT id FROM ins
+	UNION  ALL
+	SELECT id FROM tickets
+	WHERE  tx_hash = $1 AND block_hash = $2
+	LIMIT  1;`
+
+	SelectTicketsInBlock         = `SELECT * FROM tickets WHERE block_hash = $1;`
+	SelectTicketsTxDbIDsInBlock  = `SELECT purchase_tx_db_id FROM tickets WHERE block_hash = $1;`
+	SelectTicketsForAddress      = `SELECT * FROM tickets WHERE stakesubmission_address = $1;`
+	SelectTicketsForPriceAtLeast = `SELECT * FROM tickets WHERE price >= $1;`
+	SelectTicketsForPriceAtMost  = `SELECT * FROM tickets WHERE price <= $1;`
+	SelectTicketIDHeightByHash   = `SELECT id, block_height FROM tickets WHERE tx_hash = $1;`
+	SelectTicketIDByHash         = `SELECT id FROM tickets WHERE tx_hash = $1;`
+	SelectUnspentTickets         = `SELECT id, tx_hash FROM tickets WHERE spend_type = 0 OR spend_type = -1;`
+
+	// Update
+	SetTicketSpendingInfoForHash = `UPDATE tickets
+		SET spend_type = $5, spend_height = $3, spend_tx_db_id = $4
+		WHERE tx_hash = $1 and block_hash = $2;`
+	SetTicketSpendingInfoForTicketDbID = `UPDATE tickets
+		SET spend_type = $4, spend_height = $2, spend_tx_db_id = $3
+		WHERE id = $1;`
+	SetTicketSpendingInfoForTxDbID = `UPDATE tickets
+		SET spend_type = $4, spend_height = $2, spend_tx_db_id = $3
+		WHERE purchase_tx_db_id = $1;`
+
+	// Index
+	IndexTicketsTableOnHashes = `CREATE UNIQUE INDEX uix_ticket_hashes_index
+		ON tickets(tx_hash, block_hash);`
+	DeindexTicketsTableOnHashes = `DROP INDEX uix_ticket_hashes_index;`
+
+	IndexTicketsTableOnTxDbID = `CREATE UNIQUE INDEX uix_ticket_ticket_db_id
+		ON tickets(purchase_tx_db_id);`
+	DeindexTicketsTableOnTxDbID = `DROP INDEX uix_ticket_ticket_db_id;`
+
+	DeleteTicketsDuplicateRows = `DELETE FROM tickets
+		WHERE id IN (SELECT id FROM (
+				SELECT id, ROW_NUMBER()
+				OVER (partition BY tx_hash, block_hash ORDER BY id) AS rnum
+				FROM tickets) t
+			WHERE t.rnum > 1);`
+
+	// Votes
+
+	CreateVotesTable = `CREATE TABLE IF NOT EXISTS votes (
+		id SERIAL PRIMARY KEY,
+		height INT4,
+		tx_hash TEXT NOT NULL,
+		block_hash TEXT NOT NULL,
+		candidate_block_hash TEXT NOT NULL,
+		version INT2,
+		vote_bits INT2,
+		block_valid BOOLEAN,
+		ticket_hash TEXT,
+		ticket_tx_db_id INT8,
+		ticket_price FLOAT8,
+		vote_reward FLOAT8
+	);`
+
+	// Insert
+	insertVoteRow0 = `INSERT INTO votes (
+		height, tx_hash,
+		block_hash, candidate_block_hash,
+		version, vote_bits, block_valid,
+		ticket_hash, ticket_tx_db_id, ticket_price, vote_reward)
+	VALUES (
+		$1, $2,
+		$3, $4,
+		$5, $6, $7,
+		$8, $9, $10, $11) `
+	insertVoteRow = insertVoteRow0 + `RETURNING id;`
+	// insertVoteRowChecked = insertVoteRow0 + `ON CONFLICT (tx_hash, block_hash) DO NOTHING RETURNING id;`
+	upsertVoteRow = insertVoteRow0 + `ON CONFLICT (tx_hash, block_hash) DO UPDATE 
+		SET tx_hash = $3, block_hash = $4 RETURNING id;`
```

Comment:
should also be a do nothing, but not for the reasons mentioned in the other comment.